source/common/upstream/cluster_manager_impl.cc

```diff
@@ -343,6 +371,68 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   }
 }
 
+bool ClusterManagerImpl::scheduleUpdate(const Cluster& cluster, uint32_t priority, bool mergeable) {
+  const auto& update_merge_window = cluster.info()->lbConfig().update_merge_window();
+  const auto timeout = DurationUtil::durationToMilliseconds(update_merge_window);
+
+  // Find pending updates for this cluster.
+  if (updates_map_.count(cluster.info()->name()) != 1) {
+    updates_map_[cluster.info()->name()] = std::make_unique<PendingUpdatesByPriorityMap>();
+  }
+  PendingUpdatesByPriorityMapPtr& updates_by_prio = updates_map_[cluster.info()->name()];
+
+  // Find pending updates for this priority.
+  if (updates_by_prio->count(priority) != 1) {
+    (*updates_by_prio)[priority] = std::make_unique<PendingUpdates>();
+  }
+  PendingUpdatesPtr& updates = (*updates_by_prio)[priority];
+
+  // Has an update_merge_window gone by since the last update? If so, don't schedule
+  // the update so it can be applied immediately. Ditto if this is not a mergeable update.
+  const auto delta = std::chrono::steady_clock::now() - updates->last_updated_;
+  const uint64_t delta_ms = std::chrono::duration_cast<std::chrono::milliseconds>(delta).count();
+  if (delta_ms > timeout || !mergeable) {
+    // If there was a pending update, we count this update as merged update.
```

Comment:
This can only happen in the !mergeable case, right? If that's true, can you add an ASSERT to that effect inside the if statement?

Reply:
hmm, what do you mean? If `delta_ms > timeout` is true, this can happen with `mergeable==true`...

IOW, if we are not inside a merge window we will deliver this immediately (don't schedule) *even* if it's a mergeable update...

Reply:
Sorry I meant the `if (updates->disableTimer()) {` case below. The timer would have fired, right? Or I guess maybe there is a race condition? I'm just trying to understand if we can return true from `disabledTimer()` in the case where delta_ms is > timeout

Reply:
Yeah it's racy. It's possible that the timer was enabled and `delta_ms > timeout`. If it happens, we'd deliver an additional update. We could avoid the additional update by checking `timer_enabled_` in `applyUpdates()`, if it's off `applyUpdates()` lost the race and shouldn't deliver an extra update.

Thoughts?

Reply:
What you have here is fine, I would just add another comment about this race case.

Reply:
Actually, I take back what I said about the extra update being delivered. If `scheduleUpdate()` wins the race, it'll successfully disarm the timer so the extra update won't happen. Life is good.